src/main/java/io/debezium/connector/yugabytedb/YugabyteDBSnapshotChangeEventSource.java

```diff
@@ -460,10 +458,48 @@ && isSnapshotCompleteMarker(OpId.from(tabletToExplicitCheckpoint.get(tabletId)))
                   }
                 }
 
-                OpId finalOpId = new OpId(resp.getTerm(), resp.getIndex(), resp.getKey(), 
+                OpId finalOpId = new OpId(resp.getTerm(), resp.getIndex(), resp.getKey(),
                                           resp.getWriteId(), resp.getSnapshotTime());
                 LOGGER.debug("Final OpId is {}", finalOpId);
-                
+
+                // If EXPLICIT checkpointing is enabled then check if the checkpoint is the marker for snapshot completion
+                // and in case it is IMPLICIT checkpointing, the marker value should be checked on the response op_id
+                // we are receiving from the server.
+
+                /*
+                   This block checks and validates for two scenarios:
+                   1. Explicit checkpointing:
+                      a. In case of explicit, check whether the checkpoint received in the callback
+                         is the checkpoint complete marker so that the snapshot can be marked as
+                         completed.
+                      b. If the explicit checkpoint is not the snapshot end marker, there is
+                         a possibility that the finalOpId received above in the response is the
+                         snapshot complete marker - add the tablet to the set so that we do not end
+                         up calling redundant GetChanges
+                   2. Implicit checkpointing:
+                      a. In this case, only checking the response final_op_id is enough to mark
+                         the snapshot as completed.
+                 */
+                if (taskContext.shouldEnableExplicitCheckpointing()) {
+                  if (isSnapshotCompleteMarker(OpId.from(this.tabletToExplicitCheckpoint.get(tabletId)))) {
```

Comment:
Why not always add the tabletId to `snapshotCompletedTablets`? In that case, the callback is a no-op which should not matter.

Reply:
That is because `snapshotCompletedTablets` contain the list of the tablets for which the following two conditions are met:
1. The server has send the snapshot end marker.
2. In case of EXPLICIT checkpointing - Kafka has sent the callback so we are sure we have received the data.

The additional set is for cases of EXPLICIT checkpointing only where point 2 is not satisfied, so we know that server has send the data (1 is satisfied) but Kafka hasn't acknowledged the message's presense. If we always add the tabletId to `snapshotCompletedTablets` - there is a chance that when the connector crashes, we may lose some data since we may not have published them to Kafka yet.

Reply:
Re: adding tests - it's a little difficult to reproduce the issue in our tests, I spent some time but perhaps not enough to keep manipulating the tests. It was getting reproduced in the stress runs easily, which is now passing too.

Reference run: http://stress.dev.yugabyte.com/stress_test/c305a9d5-1096-45b3-a3ef-0bb39aa9ec5f

Reply:
Can you add this explanation as comments in the  EXPLICIT section?

Reply:
Added.